internal/controller/biosversionset_controller.go

```diff
@@ -0,0 +1,355 @@
+			currentStatus.InProgress += 1
+		case metalv1alpha1.BIOSVersionStatePending, "":
+			currentStatus.InProgress += 1
+		}
+	}
+	return currentStatus
+}
+
+func (r *BIOSVersionSetReconciler) getOwnedBIOSVersions(
+	ctx context.Context,
+	biosVersionSet *metalv1alpha1.BIOSVersionSet,
+) (*metalv1alpha1.BIOSVersionList, error) {
+	biosVersionList := &metalv1alpha1.BIOSVersionList{}
+	if err := clientutils.ListAndFilterControlledBy(ctx, r.Client, biosVersionSet, biosVersionList); err != nil {
+		return nil, err
+	}
+	return biosVersionList, nil
+}
+
+func (r *BIOSVersionSetReconciler) getServersBySelector(
+	ctx context.Context,
+	biosVersionSet *metalv1alpha1.BIOSVersionSet,
+) (*metalv1alpha1.ServerList, error) {
+	selector, err := metav1.LabelSelectorAsSelector(&biosVersionSet.Spec.ServerSelector)
+	if err != nil {
+		return nil, err
+	}
+	serverList := &metalv1alpha1.ServerList{}
+	if err := r.List(ctx, serverList, client.MatchingLabelsSelector{Selector: selector}); err != nil {
+		return nil, err
+	}
+
+	return serverList, nil
+}
+
+func (r *BIOSVersionSetReconciler) updateStatus(
+	ctx context.Context,
+	log logr.Logger,
+	currentStatus *metalv1alpha1.BIOSVersionSetStatus,
+	biosVersionSet *metalv1alpha1.BIOSVersionSet,
+) error {
+
+	biosVersionSetBase := biosVersionSet.DeepCopy()
+
+	biosVersionSet.Status = *currentStatus
+
+	if err := r.Status().Patch(ctx, biosVersionSet, client.MergeFrom(biosVersionSetBase)); err != nil {
+		return fmt.Errorf("failed to patch BIOSVersionSet status: %w", err)
+	}
+
+	log.V(1).Info("Updated biosVersionSet state ", "new state", currentStatus)
+
+	return nil
+}
+
+func (r *BIOSVersionSetReconciler) enqueueBiosVersionSetByServersListChanges(
+	ctx context.Context,
+	obj client.Object,
+) []ctrl.Request {
+	log := ctrl.LoggerFrom(ctx)
+	host := obj.(*metalv1alpha1.Server)
+
+	// there is no change in server list (as its not deleted or created)
+	if host.DeletionTimestamp.IsZero() && controllerutil.ContainsFinalizer(host, ServerFinalizer) {
+		return nil
+	}
+
+	biosVersionSetList := &metalv1alpha1.BIOSVersionSetList{}
+	if err := r.List(ctx, biosVersionSetList); err != nil {
+		log.Error(err, "failed to list BIOSVersionSet")
+		return nil
+	}
+
+	for _, biosVersionSet := range biosVersionSetList.Items {
+		serverList, err := r.getServersBySelector(ctx, &biosVersionSet)
+		if err != nil {
+			log.Error(err, "failed to list BIOSVersionSet servers by selector")
+			return nil
+		}
+		for _, server := range serverList.Items {
+			if server.Name == host.Name {
+				ownedBiosVersion, err := r.getOwnedBIOSVersions(ctx, &biosVersionSet)
+				if err != nil {
+					log.Error(err, "failed to list owned BIOSVersions")
+					return nil
+				}
+				if len(ownedBiosVersion.Items) != len(serverList.Items) {
+					return []ctrl.Request{{
+						NamespacedName: client.ObjectKey{
+							Name:      biosVersionSet.Name,
+							Namespace: biosVersionSet.Namespace,
+						}}}
+				}
+			}
+		}
+	}
+	return nil
```

Comment:
```suggestion
	reqs := make([]ctrl.Request, 0)
	for _, biosVersionSet := range biosVersionSetList.Items {
		selector, err := metav1.LabelSelectorAsSelector(&biosVersionSet.Spec.ServerSelector)
		if err != nil {
			log.Error(err, "failed to convert label selector")
			return nil
		}
		if selector.Matches(labels.Set(host.GetLabels())) {
			reqs = append(reqs, ctrl.Request{
				NamespacedName: client.ObjectKey{
					Name:      biosVersionSet.Name,
					Namespace: biosVersionSet.Namespace,
				},
			})
		}
	}
	return reqs
}
```
When calling `getServersBySelector` you list Servers but later you simply filter only host from them. It is much easier to simply check if host matches label selector. I also don't quite get why you added `if len(ownedBiosVersion.Items) != len(serverList.Items)`. I assume to prevent reqonciling without changes but it shouldn't be done like this. One more reconcile without change doesn't make a different but to save some extra work we can add a pridicate to watch. Please check my next comment.